### DISCUSSION TITLE
Seeing 'ERROR HERE' during tracking step

### DISCUSSION THREAD

**Discussion by hannahwalk** [ronshnapp/MyPTV#36] [Q&A]
MyPTV is matching the particles between my images beautifully and I can see clear patterns in the flow just using the matched particles but when it comes to tracking I'm having some issues. 
It's printing 'ERROR HERE' (from tracking_mod.py in a try/except ~line 240) and then a list of particle positions many times (either every frame or every particle it is tracking--it's hard to tell how many times it has printed). 

At the end I still get some trajectories but after smoothing and stitching they are sparse and jumpy. I have ~40 particles being matched per image and then end up with < 10 bad trajectories per image.

Does the 'ERROR HERE' statement give any idea of something that is going wrong in the tracking based on parameters I've input?
Could I have too many particles that I'm matching which makes it hard to find the nearest neighbor?

Thank you!

**Comment by ronshnapp** [comment]
Hi Hannah,

I'm not sure exactly what is causing this error. It's probably not because of the number of particles you're matching per frame as MyPTV can handle thousands of matched particles per frame. But just to be sure, maybe it's worth while trying to redo some of the analysis using other parameters.
 
If you could perhaps share one of your matched particles files and the params file, I'll try to redo the tracking on my machine and trace back the error and see whether it's a bug or something else.

Best,
Ron

**Reply by hannahwalk** [reply]
Hi Ron,
Thank you for the quick reply!
I'll continue adjusting my tracking parameters to see if that helps.
I also have accumulated relevant files in case you have time to take a look. I wrote a very brief description of the issue in it as well and am very happy to provide more information. 
It matches all of the particles well and in the predicted geometry of the camera sampling volume. 
Your help is greatly appreciated. 
Thank you so much!
Hannah
[myPTV_HW_files.zip](https://github.com/ronshnapp/MyPTV/files/11693335/myPTV_HW_files.zip)

**Reply by ronshnapp** [reply]
Hi Hannah,

Thanks for sending these files!
I was looking at the data you sent. I agree that it seems the matching went fine. In the segmentation file I can spot "streaks" of dots however it's hard for me to "mentally see" which dots I should by connected in time. Therefore, I think that the problem might be related to the acquisition rate in the experiment. In order to be able to "connect the dots" in the tracking step, we usually apply the "trackability" condition meaning that the distances that particles traverse in between two consecutive frames is much smaller than the typical distance between particles. See the movie I sent of an example dataset I used to make the example folder in MyPTV. In your data, there are about 150 particles in each frame,  so with the resolution used I would try to adjust the framerate to make sure particles move up to about 5 pixels in between each frame. Do you think this might solve the problem?



![segmentation_frame0_4](https://github.com/ronshnapp/MyPTV/assets/34024910/84847c67-c6f9-4414-abf4-d9ab785e0cc1)

https://github.com/ronshnapp/MyPTV/assets/34024910/d9826413-c49c-448e-8627-47c04cfe5221

Best,
Ron

**Reply by hannahwalk** [reply]
Ah yes, I'm seeing the same thing. That's unfortunate. 
I was wondering if you had thought about using a basic PIV to guide the particle tracking search. I've done this using 2D PTV before to help improve and speed up tracking process and it worked well. It allowed me to find particles in this same data set that were far apart (up to 60 pixels apart).  
I am hoping to add that into my usage of myPTV but wanted to hear your thoughts.

**Comment by ronshnapp** [comment]
I have to say I've never tried something like that, and that it sounds a cool idea. For 3D-PTV, tough, one would need to do the correlation using voxels rather then pixels, something like tomographic PIV. Overall it's a cool idea, and if you think you'd like to try and develop this I'd be very glad to help. We could zoom about this in case you'e interested. 
But in another direction that could help, when I first started writing the tracking module it had the option of specifying  a "mean velocity", such that if there's a flow with a strong mean translation component it could make tracking easier. The problem is I didn't add this feature to the params file and workflow script. If you think this is something that could be useful for you, I could show you how to use this option, either by running the tracking manually (i.e. using the module without the workflow script) or by adding this feature to the workflow.

Let me know your thoughts.

Best,
Ron

**Reply by hannahwalk** [reply]
A zoom chat would be great! 
Yes, I did see that part in the manual and was hoping to get the mean flow component implemented in my tracking. 
I'm attending and presenting at ISPIV next week (a conference you may be attending?) and am very interested in chatting once that is over.

**Reply by ronshnapp** [reply]
Unfortunately I will not be at the ISPIV this year, although it is a great conference, so good luck with your presentation! While you're there, say hi to @alexlib who is giving a keynote keynote lecture there.
About the Zoom - will the 26th or 27th of June work for you?

**Reply by hannahwalk** [reply]
Yes I will!
Yes, both dates work for me. My email is h2walker@ucsd.edu.
